Fix screen getting stuck after #9 fix

vladkorotnev opened the pull request:
As mentioned in https://github.com/Finnerale/druid-enums/issues/8#issuecomment-801101934, the #9 fix made the screen freeze when returning to an older state with the same data due to the missing `children_changed` call.

vladkorotnev committed:
Fix screen getting stuck after #9 fix

Finnerale commented:
Thanks for the proposed fix, but calling `children_changed` unconditionally is not the right way to handle this.
Instead, we should call `ctx.request_paint();` where you are calling `childaren_changed` right now, in addition to `false => ctx.children_changed()`.
Your fix also worked, because `children_changed` calls `request_layout`  and that causes `request_paint`. Just calling `request_paint` is enough though.

vladkorotnev committed:
Pull-request review corrections

Calling children_changed is overkill because the actual hierarchy didn't change, but as a side effect it called request_paint which fixes the bug — so call that directly instead.

vladkorotnev committed:
Now that just crashed everything

vladkorotnev committed:
Update lib.rs

vladkorotnev committed:
Revert fixes

vladkorotnev commented:
I tried a few variants of calling `request_paint` or `request_layout` but sadly they all crashed at

```
thread 'main' panicked at 'WidgetId(48): layout method called before receiving WidgetAdded.', C:\Users\akasaka\.cargo\git\checkouts\druid-f71533f3d81c0bc8\17e03bd\druid\src\core.rs:535:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

even when only calling request_layout from under the `true` branch of the widget initialization state check. 

Guess I'll resort to using my fork for now and either investigate further someday later or have to leave it to you :D

Finnerale commented:
Huh, this is what I had in mind:
```rust
                ctx.request_paint();
                match &mut self.#builder_name {
                    Some(widget) => match widget.is_initialized() {
                        true => widget.update(ctx, #data_values, env),
                        false => ctx.children_changed(),
                    },
                    None => (),
                }
```
Does that work for you, or do you still get errors with it?

vladkorotnev committed:
Fix as suggested in review (2)

vladkorotnev commented:
Just tried it now, still freezing

Finnerale commented:
Odd, it works for me with the example you provided. And you did add the `request_paint` call when testing it? Because it's missing in this PR right now.

vladkorotnev committed:
Add lost request_paint call

vladkorotnev committed:
And yet another typo of course

vladkorotnev commented:
Sorry, I haven't slept for a while now and now this pull request is very awkward with so many commits for just a few lines... I blame Cargo for not having support for both `path` and `git` in a dependency at the same time xD
 
Yep, with the `request_paint` it works fine now! Thanks for your help once more.

Finnerale commented:
Glad it works now. I'll just squash and merge this :)
